parallel netcdf IO

jswhit2 opened the pull request:
this enables parallel IO for netcdf output - set 'netcdf_parallel' for output_file in model_configure.  Works for compressed writes.

~~TODO:  add ifdef's around parallel code sections so parallel versions of netcdf/hdf5 will not be required~~

To compile without netcdf parallel support, use `-DNO_PARALLEL_NETCDF`

https://github.com/NOAA-EMC/fv3atm/issues/23

Supersedes https://github.com/NOAA-EMC/fv3atm/pull/19

junwang-noaa committed:
netcdf parallel writing and lon/lat in netcdf file

Commit:
some changes to get it to (almost) run on hera

Commit:
lat should be in degrees - remove conversion to radians.

Commit:
import updates from jswhit/fv3atm

Commit:
more bug fixes - now works on hera

Commit:
specify collective access if compression turned on.

Commit:
use classic model

Commit:
bug fixes for parallel IO with compression

Commit:
fix calculation of max compression error

Commit:
turn off shuffle filter

Commit:
code simplification

Commit:
remove debug print

Commit:
don't use parallel IO for 2d file (since it seems to increase run time)

Commit:
allow multiple values of output_file, as long as they all start with 'netcdf'

Commit:
use default chunksize for 2d vars

DusanJovic-NOAA commented:
rad2dg can be deleted from module_write_netcdf.F90

DusanJovic-NOAA commented:
@junwang-noaa Please delete ESMF_ERR_ABORT define (first 3 lines in a file) from module_write_netcdf.F90 and module_write_netcdf_parallel.F90. This macro is not used.

Commit:
delete commented out macro ESMF_ERR_ABORT

Commit:
delete rad2dg

Commit:
add module_write_netcdf_parallel.F90

Commit:
add option to build without parallel netcdf (-DNO_PARALLEL_NETCDF)

Commit:
fix typo

Commit:
stub file for building without parallel netcdf lib

Commit:
Merge remote-tracking branch 'upstream/develop' into netcdf_parallel

Commit:
allow chunksizes for 2d arrays to be set in model_configure (ichunk2d,jchunk2d)
Default is size of array on each write task.

edwardhartnett commented on io/module_write_netcdf_parallel.F90:
```diff
@@ -171,7 +172,7 @@ subroutine write_netcdf_parallel(fieldbundle, wrtfb, filename, mpi_comm, mype, i
                     (/im_dimid,jm_dimid,pfull_dimid,time_dimid/), varids(i), &
                     shuffle=.false.,deflate_level=ideflate,&
-                    chunksizes=(/im,jm,1,1/),cache_size=40*im*jm); NC_ERR_STOP(ncerr)
+                    chunksizes=(/im,jm,1,1/)); NC_ERR_STOP(ncerr)
```
Must two of the chunksizes be 1?

jswhit2 replied:
No, but this seems to be the most common access pattern (reading a 2d horizontal slice).

jswhit replied:
Now changed so that chunksize can be specified at run time.  Default is to use MPI decompisition size..

Commit:
add ichunk3d,jchunk3d,kchunk3d to specify 3d chunksizes.
Default is now ichunk3d,jchunk3d same as array size on each PE, kchunk3d=nlevs
This results in the fastest writes on hera.

DusanJovic-NOAA commented on io/CMakeLists.txt:
```diff
@@ -5,13 +5,21 @@ else()
   add_definitions(-DNO_INLINE_POST)
 endif()
 
+if(PARALLEL_NETCDF)
```
file names must be module_write_* not model_write_*

jswhit replied:
FIxed, thanks

DusanJovic-NOAA commented on io/CMakeLists.txt:
```diff
@@ -5,13 +5,21 @@ else()
 endif()
 
+if(PARALLEL_NETCDF)
+  set(NETCDF_PARALLEL_SRC model_write_netcdf_parallel.F90)
```
Instead of maintaining separate stub file and maintaining changes in CMakeLists.txt and makefile wouldn't be easier to change module_write_netcdf_parallel.F90 to this structure (or something similar):
```
module module_write_netcdf_parallel

 ....

  private                                                                       
    public write_netcdf_parallel                                                  
                                                                                  
    contains                                                                      
                                                                                  
#ifndef NO_PARALLEL_NETCDF 

  <current content of module_write_netcdf_parallel.F90 subroutines>

#else

    subroutine write_netcdf_parallel(fieldbundle, wrtfb, filename, mpi_comm, mype, im, jm, rc)

     type(ESMF_FieldBundle), intent(in) :: fieldbundle
     type(ESMF_FieldBundle), intent(in) :: wrtfb
     character(*), intent(in)           :: filename
     integer, intent(in)                :: mpi_comm
     integer, intent(in)                :: mype
     integer, intent(in)                :: im, jm
     integer, optional,intent(out)      :: rc

     print *,'in write_netcdf_parallel - model not built with parallel netcdf support, return'

   end subroutine write_netcdf_parallel

#endif

end module module_write_netcdf_parallel
```

jswhit replied:
Good idea - done.

Commit:
fix typo

Commit:
put ifdefs in module_write_netcdf_parallel.F90 so no stub file needed

Commit:
don't need this file anymore

Commit:
remove module_write_netcdf_parallel_stub.o target

Commit:
use specified chunksizes for serial IO. If chunksize parameter negative, let netcdf library choose defaults.

Commit:
update comments

junwang-noaa committed:
get output_file without esmf call error

junwang-noaa committed:
syntax fix

junwang-noaa committed:
fix stub interface in module_write_netcdf_parallel.F90 for cmake build

edwardhartnett reviewed (commented):
The netcdf parts look OK.

Are there no tests in this repo? Where do you guys do testing?

junwang-noaa commented:
We are doing the tests with FV3 parallel code. Currently we are in the
process of updating regression test. Will talk to you in next weeks on
setting up CI.

Jun

On Tue, Feb 4, 2020 at 1:08 PM Edward Hartnett <notifications@github.com>
wrote:

> *@edwardhartnett* commented on this pull request.
>
> The netcdf parts look OK.
>
> Are there no tests in this repo? Where do you guys do testing?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/fv3atm/pull/41?email_source=notifications&email_token=AI7D6TPW5ZBHDJ5MQC6YAH3RBGVKHA5CNFSM4KFKDGJ2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCUGT5RY#pullrequestreview-353189575>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI7D6TM7Y4VKXNYOQGEUS3DRBGVKHANCNFSM4KFKDGJQ>
> .
>

jswhit commented:
I am OK with this PR as is.

junwang-noaa commented:
Thanks, Jeff and Ed for reviewing the code.

On Tue, Feb 4, 2020 at 1:46 PM Jeff Whitaker <notifications@github.com>
wrote:

> I am OK with this PR as is.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/fv3atm/pull/41?email_source=notifications&email_token=AI7D6TLV7EG5SMPFKW7RGHDRBGZXRA5CNFSM4KFKDGJ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKYXZOI#issuecomment-582057145>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI7D6TOOEM4LLVU72637VCDRBGZXRANCNFSM4KFKDGJQ>
> .
>
